pages/docs/evolution/breaking-changes.md

```diff
@@ -0,0 +1,483 @@
+
+## `3` Types of incompatible changes and corresponding guidelines
+
+`3.1` Small fixes that virtually no users will
+    encounter can normally be made right away (still require committee
+    [review](#1.4)).
+
+* `3.1.1` When unsure about the full risk/impact of
+    a change:
+
+*   Implement the change in a preview build.
+*   Try on large bodies of code available to us.
+*   Fall back to [deprecation procedure](#4) if needed.
+
+`3.2` Behavior-changing bug fixes in `kotlinc`:
+
+* `3.2.1` Good code did not compile due to a bug can be
+    fixed right away.
+
+* `3.2.2` Bad code compiled due to a bug, but always
+    failed at runtime can be fixed right away.
+
+* `3.2.3` Bad code compiled due to a bug, and worked
+    reasonably, must follow the deprecation policy spelled out in
+    [Paragraph 4](#4).
+
+
+`3.3` Bug fixes and contract refinement in the
+ `kotlin-stdlib` require publishing a release note one version in advance.
+
+`3.4` Retiring language features & `kotlin-stdlib` APIs
+    must follow the deprecation policies spelled out in
+    [Paragraph 4](#4) and [Paragraph 5](#5).
+
+
+
+## `4` Deprecation procedures for language features
+
+Deprecation is closely related to versioning, for versioning scheme
+we refer to the existing [versioning policy](../reference/compatibility.html).
+
+`4.1` Before any changes are made, the upcoming
+    change along with its rationale and the deprecation/migration plan
+    need to be announced to the Kotlin community, 
+    preview builds of the compiler and standard library should be made available.
+
+`4.2` Release a version (call it N.M) that reports
+    deprecation warnings for the feature:
+
+* `4.2.1` Provide and describe a replacement for
+    deprecated constructs as part of the warning.
+
+* `4.2.2` Allow to disable the warnings in settings,
+    or promote them to errors.
+
+* `4.2.3` Along with this version, provide automated
+    migration aids (e.g. in the IDE).
+
+`4.3` Allow a long enough period of time for
+    migration, the exact time will be based on the impact and complexity
+    of the change as discussed during the Kotlin Language Committee
+    [review](change-review-process.html).
+
+`4.4` In the version N.M+1 or N+1, report errors for
+ previously deprecated cases.
+
+* `4.4.1` In the release notes, repeat the information
+    about the deprecation plan.
+
+* `4.4.2` Allow to demote the errors to warnings through
+    settings.
+
+* `4.4.3` Keep migration aids available.
+
+`4.5` In the version N.M+2 or N+1, report errors for
+    the deprecated feature.
+
+* `4.5.1` In the release notes, declare the feature as
+    discontinued.
+
+* `4.5.2` If possible, keep a compatibility mode that
+    does not support new features, but allows the deprecated one as in
+    version N.M+1 or N+1.X.
+
+* `4.5.3` Migration aids can be removed from the
+    compiler at this point.
+
+`4.6` A version that supports automated migration
```

Comment:
It is unclear which tools are covered by this paragraph. Since it mentions "automated migration", I assume that it includes not only the compiler, but also the IDE and the IDE plugin. In this case, it is unclear to me how it will work. Will JetBrains support installing older versions of the Kotlin plugin? Will the release notes contain "upgrade with x - y plugin version" notices? 

As a solution I can propose extracting the language migration functionality into separate plugins. Here is an example of how it would work: suppose I want to migrate my code base from Kotlin 1.6 to Kotlin 1.8; then I install the latest version of the Kotlin plugin, plus two smaller plugins: "Kotlin Migrator 1.6 -> 1.7" and "Kotlin Migrator 1.7 -> 1.8", where the additional plugins provide inspections and quick fixes needed to migrate from 1.6 to 1.7, and from 1.7 to 1.8 respectively. This configuration allows for different update cycles for the main plugin and for the migration assisting plugins. The downside is that it would probably require a somewhat stable internal compiler API to allow the plugins to communicate.

Reply:
That's a good option